src/TokenAirdrop.hs

```diff
@@ -66,27 +70,41 @@ tokenAirdrop config = do
           putStrLn "Answer is not valid"
           confirmTxSubmission
 
+    handle' :: String -> IO (Either Text ()) -> IO (Either Text ())
+    handle' s m = catch @SomeException m $ \e -> do
+      when config.verbose $ print e
+      pure . Left . pack $ s
+
     processTransactions :: [(Constraints.TxConstraints Void Void, [Beneficiary], Int)] -> Map PubKeyHash PubKeyAddress -> IO (Either Text ())
-    processTransactions txs pubKeyAddressMap =
-      mapMErr
-        ( \(tx, bs, i) -> do
-            putStrLn $ "Preparing transaction " ++ show i ++ " of " ++ show (length txs) ++ " for following benficiaries:"
-            mapM_ print bs
-
-            utxos <- utxosAt config $ config.ownAddress
-            let lookups = Constraints.unspentOutputs utxos
-
-            eTxId <- submitTx @Void config pubKeyAddressMap lookups tx
-            case eTxId of
-              Left err -> pure $ Left err
-              Right txId -> do
-                when config.live $ do
-                  putStrLn $ "Submitted transaction successfully: " ++ show txId
-                  putStrLn "Waiting for confirmation..."
-                  waitUntilHasTxIn config 0 txId
-                pure $ Right ()
-        )
-        txs
+    processTransactions txs pubKeyAddressMap = do
+      result <- flip mapMErr txs $
+        \(tx, bs, i) -> handle' ("Failed to prepare and submit transaction " ++ show i) $ do
+          putStrLn $ "Preparing transaction " ++ show i ++ " of " ++ show (length txs) ++ " for following benficiaries:"
+          mapM_ print bs
+
+          utxos <- utxosAt config $ config.ownAddress
+          let lookups = Constraints.unspentOutputs utxos
+
+          eTxId <- submitTx @Void config pubKeyAddressMap lookups tx
+          case eTxId of
+            Left err -> pure $ Left err
+            Right txId -> handle' ("Failed to confirm transaction: " ++ show txId) $ do
+              when config.live $ do
+                putStrLn $ "Submitted transaction successfully: " ++ show txId
+                putStrLn "Waiting for confirmation..."
+                waitUntilHasTxIn config 0 txId
+              pure $ Right ()
+
+      case result of
+        Right _ -> pure $ Right ()
+        Left (msg, failed NEL.:| remaining) -> do
+          when config.live $ do
+            let showBeneficiaries (_, bens, _) = fmap show bens
+            createDirectoryIfMissing True $ takeDirectory config.currentBeneficiariesLog
+            writeFile config.currentBeneficiariesLog . unlines $ showBeneficiaries failed
+            createDirectoryIfMissing True $ takeDirectory config.remainingBeneficiariesLog
+            writeFile config.remainingBeneficiariesLog . unlines $ remaining >>= showBeneficiaries
+          pure $ Left msg
```

Comment:
This whole function has gotten pretty large now, any way we can split it up a bit?